### ISSUE TITLE
A ServiceEntry from one namespace can interfere with the traffic of another namespace

### ISSUE THREAD

**Issue by uk000** [istio/istio#13008]
**Describe the bug**
As of Istio 1.1.1, anybody can create a ServiceEntry in one's own namespace to interfere with or hijack traffic of all other namespaces for a certain host.
The hijacked traffic can be routed to another external destination, or to some internal kubernetes service. Even if we ignore the possibility of a malicious intent, this leaves the door open for a simple mistake to disrupt production traffic for crucial applications.

In some ways, this issue is in the same league as #12144 which was in the context of DestinationRule interfering with unrelated namespaces.

**Expected behavior**
A ServiceEntry should only take effect in the namespace where it is defined, and not apply to all namespaces in the cluster. That way, whether by malicious intent or honest mistake, a user can only disrupt traffic of one's own namespace and not for the whole cluster.

**Steps to reproduce the bug**
1. Given two namespaces client-ns-1 and client-ns-2, both running some client deployment that provides curl (e.g. sleep or testcurl).
Also given there's an external application accessible at some fqdn or IP address

Initially neither clients can access kubeman.com.

2. Create a ServiceEntry in client-ns-1 to enable access to the external service

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: kubeman-ext
spec:
  exportTo:
    - "."
  hosts:
  - kubeman.com
  ports:
  - number: 80
    name: http-kubeman-80
    protocol: HTTP
  location: MESH_EXTERNAL
  resolution: DNS
  endpoints:
  - address: kubeman.com
  ```

Now verify that client-ns-1 can access the external service. Note that this client-1 created this ServiceEntry for its own local namespace. Other possible variations in this step: 
- Use STATIC resolution with IP address instead of DNS
- export ServiceEntry to all namespaces instead of client-ns-1 only  

3. Now client-2 creates a ServiceEntry to route domain kubeman.com to somewhere else, basically hijacking the domain

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: kubeman-to-google
spec:
  hosts:
  - kubeman.com
  ports:
  - number: 80
    name: http-kubeman-80
    protocol: HTTP
  location: MESH_EXTERNAL
  resolution: DNS
  endpoints:
  - address: google.com
  ```

Now the requests from any namespace to kubeman.com are randomly routed to either kubeman.com or google.com

4. Alternately, traffic can be hijacked to route to an internal kubernetes service

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: kubeman-int
spec:
  hosts:
  - kubeman.com
  ports:
  - number: 80
    name: http-kubeman-80
    protocol: HTTP
  location: MESH_INTERNAL
  resolution: DNS
  endpoints:
  - address: nginx.svc-sc-nomtls-1.svc.cluster.local
```

5. Additionally, traffic can be disrupted for an internal service if a rogue ServiceEntry gets created hijacking the internal service FQDN. In this case the traffic doesn't get routed to the new destination, but seems to be simply disrupted with envoy returning 503 "UH".
To achieve this, create a ServiceEntry using the internal service fqdn as "host" and attempt to send the traffic elsewhere:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: good-to-bad
spec:
  hosts:
  - good-service.good-namespace.svc.cluster.local
  ports:
  - number: 80
    name: http-80
    protocol: HTTP
  location: MESH_INTERNAL
  resolution: DNS
  endpoints:
  - address: bad-service.bad-namespace.svc.cluster.local
    weight: 100
```

Overall, what I found strange in my testing is that the behavior is not consistent. It seems like a case of eventual consistency or last-write-wins, or some other logic is in play. Sometimes the creation of conflicting ServiceEntry hijacks traffic completely, whereas other times it only hijacks it randomly (e.g. 50-50 split if no weights are specified in the ServiceEntry). And then at times the behavior is that it doesn't hijack traffic for the existing pods, but does hijack it if either a pod is rescheduled or if the original (first) ServiceEntry is redeployed for some reason (e.g. some changes in the first ServiceEntry). Upon redeployment, suddenly the 2nd ServiceEntry takes over and hijacks traffic.

**Version**

```
client version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.0-17-g2b13318"}
citadel version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.0-17-g2b13318"}
galley version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.0-17-g2b13318"}
ingressgateway version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.0-17-g2b13318"}
nodeagent version: version.BuildInfo{Version:"", GitRevision:"", User:"", Host:"", GolangVersion:"", DockerHub:"", BuildStatus:"", GitTag:""}
pilot version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.0-17-g2b13318"}
policy version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.0-17-g2b13318"}
sidecar-injector version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.0-17-g2b13318"}
telemetry version: version.BuildInfo{Version:"1.1.1", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"7077232d-4c6c-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.0-17-g2b13318"}
```

```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.11", GitCommit:"637c7e288581ee40ab4ca210618a89a555b6e7e9", GitTreeState:"clean", BuildDate:"2018-11-26T14:38:32Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.11", GitCommit:"637c7e288581ee40ab4ca210618a89a555b6e7e9", GitTreeState:"clean", BuildDate:"2018-11-26T14:25:46Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Installation**
Using helm template

**Environment**
kubernetes on docker-for-mac

**Cluster state**

**Comment by uk000** [comment]
Updated the bug description to add another scenario (step #5), where a ServiceEntry can disrupt traffic going to an internal service.

**Comment by costinm** [comment]
Sidecar is required to enable isolation - for backward compat we didn't enable isolation if Sidecar is not used ( since we don't know which namespace to use )

**Comment by uk000** [comment]
The behavior was the same with or without using Sidecar resource. TBH, I don't fully understand why would a sidecar resource make a difference with this issue. Isn't the sidecar resource primarily meant to restrict/allow egress routes to specific hosts? Sidecar doesn't really affect the destinations of routes.

**Comment by howardjohn** [comment]
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: httpbin-a
  namespace: a
spec:
  hosts:
  - httpbin.org
  ports:
  - number: 80
    name: http
    protocol: HTTP
  resolution: DNS
  location: MESH_EXTERNAL
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: wikipedia-b
  namespace: b
spec:
  hosts:
  - wikipedia.org
  ports:
  - number: 80
    name: http
    protocol: HTTP
  resolution: DNS
  location: MESH_EXTERNAL---
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: sidecar
spec:
  egress:
    - hosts:
      - ./*
      - istio-system/*
---
```

`istioctl pc clusters` shows no wikipedia entry.


But, we change it to the interfering with traffic like in description

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: wikipedia-b
  namespace: b
spec:
  hosts:
  - httpbin.org
  ports:
  - number: 80
    name: http
    protocol: HTTP
  resolution: DNS
  location: MESH_EXTERNAL
  endpoints:
  - address: wikipedia.org
```

And now I see routing to wikipedia and httpbin.

`istioctl pc clusters` shows:
```json
[
    {
        "name": "outbound|80||httpbin.org",
        "type": "STRICT_DNS",
        "connectTimeout": "1s",
        "loadAssignment": {
            "clusterName": "outbound|80||httpbin.org",
            "endpoints": [
                {
                    "lbEndpoints": [
                        {
                            "endpoint": {
                                "address": {
                                    "socketAddress": {
                                        "address": "wikipedia.org",
                                        "portValue": 80
                                    }
                                }
                            },
                            "loadBalancingWeight": 1
                        },
                        {
                            "endpoint": {
                                "address": {
                                    "socketAddress": {
                                        "address": "httpbin.org",
                                        "portValue": 80
                                    }
                                }
                            },
                            "loadBalancingWeight": 1
                        }
                    ]
                }
            ]
        },
        "circuitBreakers": {
            "thresholds": [
                {
                    "maxRetries": 1024
                }
            ]
        },
        "dnsRefreshRate": "5s",
        "dnsLookupFamily": "V4_ONLY"
    }
]
```

Seems to be a clear bug. @costinm

**Comment by hzxuzhonghu** [comment]
Isn't this expected, pilot merges the two service entries?

**Comment by howardjohn** [comment]
No we should not get services from other namespaces if we have Sidecar, for all the reasons mentioned in the issue.

**Comment by howardjohn** [comment]
Think I know what is happening. First we get all of the services exposed to the sidecar. Then, for each service, we look up instances by hostname. At this point, we can pull in entries from other namespaces. 
 See: https://github.com/istio/istio/blob/fdd6444d166e1c12e6ab48cffae8110b64ea26e6/pilot/pkg/networking/core/v1alpha3/cluster.go#L313

Seems like we would need pretty substantial changes in order to fix this. @rshriram any ideas?

**Comment by rshriram** [comment]
Few things:
 1. you can restrict the visibility of a service entry by declaring the service entry's `exportTo: '.'` . This will ensure that the service entry is visible only in the declaring namespace and not in others.
 2. Even when you do this exportTo stuff, you cannot have two service entries in the system with the same host. This is equivalent to having two google.coms in the service registry where each google.com points to something completely different. Such a thing is not possible in Istio today because the service registry is a map[hostnameFQDN->service]. Bad things will happen when you do this. 
 If you think about this, almost every service registry out there today (consul, zookeeper, eureka, etc.) all enforce uniqueness on the name lookup. Even inside kubernetes, you cannot have foo.ns1.svc.cluster.local inside ns1 and ns2 namespaces. k8s simply wont allow that by virtue of appending the unique namespace to the FQDN.

Perhaps what can be done is to add an additional check in the aggregate service registry and in external service registry (one processing service entries) to check for duplicates, and emit warnings.

**Comment by howardjohn** [comment]
Me and Shriram discussed this a bit offline. Adding the conversation here for others to comment.

Key points:
* My thoughts are that Sidecar should provide config isolation. In this case, we are pulling in config from other namespaces, which breaks this isolation. I think this is important because it limits our multi tenancy support.
* Shriram thinks this is expected because it breaks a fundamental concept that services can only be in one namespace. Violating this will break telemetry (not just mixer, but other systems as well like Datadog). We should only support multi-tenancy within the confines of basic concepts. Adding new abstractions for service registry will lead to too much complexity. Service entries should not be allowed to be created by teams, they are not routing related.
* Possible workaround is to add a validation rule to reject duplicate hosts.

cc @costinm @louiscryan

**Comment by costinm** [comment]
'exportTo' is not an option for security reasons. No namespace that is not explicitly imported should be able to affect the config of another namespace - clear security reasons. 

ServiceEntry is used for egress - so needs to be per-namespace

**Comment by howardjohn** [comment]
@costinm exportTo does not fix this either anyways, they will still be imported from other namespaces even if they have `exportTo: '.'`

**Comment by costinm** [comment]
Even if it fixes it (and it could be used) - it is still a security problem, because a namespace can interfere with a different one and inject an egress.

We can restrict the ServiceEntry for "foo.bar.cluster.local" to only be accepted in "bar" namespace - 
i.e. service can only be defined in its own namespace, like in K8S. 
But the egress ( www.example.com ) should not be picked from a random namespace that is not 
imported.

**Comment by costinm** [comment]
Also: validation rule doesn't work either - it is legitimate for 2 namespaces to define their own view of an egress.

**Comment by louiscryan** [comment]
@rshriram I disagree a bit with your points above.

We introduced namespace scoping of service references to avoid collisions like this. That feature also explicitly allows for split horizons across namespaces so service resolution can't simply be

map[hostnameFQDN->service]

its really map[hosnameFQDN, namespace-> service]

otherwise we'd have to treat services in the same namespace equivalently with services in other namespace when consuming them.  The old behavior is because we by default treat service references as 

*/host-name

but that isn't safe for large deployments. We need to respect scoped references otherwise they won't meet user expectations

**Comment by rshriram** [comment]
what happens if I declare a serviceEntry in my local namespace like this:
```
ServiceEntry:
 name: hack
 namespace: mybroken-namespace
host: someBigService.namespaceIdontHaveaccess.svc.cluster.local
address: 127.255.1.3 or my pod IP
ports:
  same ports as someBigService
```

such that I can hijack all traffic bound for this someBigService by declaring a namespace local service entry, and start discovering what traffic is being sent to that service?

Today, this wont work (mostly) in our system because this service is always keyed by the hostname and the hostname will already exist in k8s service registry, such that the service specs from another namespace will be ignored.

**Comment by howardjohn** [comment]
@rshriram this will work today, that is the  issue we are seeing.  This is example 5 in the original issue. The spec from a different namespace isn't ignored they get merged. (note I don't think I tried this exact scenario, so I'm not 100% sure, but I'll try to reproduce it later)

**Comment by rshriram** [comment]
We can prevent this easily by detecting and error-ing out on collisions. Preferring services from the registry over services from the config. Note my example has k8s internal service..

**Comment by louiscryan** [comment]
If a namespace owner chooses to declare a serviceentry to capture their own traffic that seems absolutely fine, they control the apps in their own namespace after all and could achieve the identical effect in other ways

Its when another namespace can cause this effect that we have an issue.

**Comment by louiscryan** [comment]
Is there any objection to a resolution mechanism of:

1. if 'host' reference is of form '<namespace>/<host>' then only serviceentry/service from that namespace will be used. Multiple entries from that namespace with the same hostname are merged

2. if 'host' reference is of form '<host>' then we treat it as equivalent to '*/<host>' with the following precedence for resolution
 - look in current namespace for 'host', if found use that
 - look in root namespace for 'host', if found use that
 - look in all namespaces for 'host', if found use one and only one namespaces definition

3. if 'host' reference is of form './<host>' then we resolve in current namespace only

Note that 2 has some potential for ambiguity if resolution falls through to the third step but the import was explicitly ambiguous.

**Comment by costinm** [comment]
I don't think we can use prefixes or other forms - the hosts are originally from routes or directly from the app calling a DNS name. 

Also - I don't think the example Shriram provided is valid or should work - it absolutely must fail if TLS is enabled ( secure naming should prevent it ). Any workload must be associated with a cert 
that has a namespace in it, and the declaration of that service must match the namespace. There is no technical reason to mess with the security semantics just to enable some hacks. 

 For external hosts ( i.e. not .cluster.local, and not subject to SIFFE ) - we already have very clear 
rules in Sidecar - which match (2) if I understand it correctly.

**Comment by rshriram** [comment]
Hmm. I didn’t think of the mtls part. Okay, so it seems there is no security issue with multiple service entries. Let’s go ahead then.

**Comment by rshriram** [comment]
My second main issue is how vms will be identified. If we assume that it’s not possible to add this config_namespace metadata to every one of the 1000s of vms, then when a proxy connects to pilot, what will getProxyServiceInstances return? (I don’t want to hear a “use certificates answer” as we don’t have that support today, and encoding some istio namespace into pre issued certs is a no go in many cases).

**Comment by hzxuzhonghu** [comment]
come up with a idea, how about prevent the conflicting rules from being created by Galley?

**Comment by sdake** [comment]
@hzxuzhonghu galley validation seems reasonable - in that unexpected behavior wouldn't occur - however - would be nice to fix the root of the problem.

Cheers
-steve

**Comment by howardjohn** [comment]
I don't think we can do galley validation of this. AFAIK galley does not support stateful validation, so we can't see there is a conflicting rule

**Comment by hzxuzhonghu** [comment]
This is a direction, that can be tried i think. As some kubernetes in-tree admission controllers also do such similar thing.

**Comment by howardjohn** [comment]
@ozevren had some interesting points of problems with making Galley support these kind of stateful validations and you had some good points but I cannot find where the conversation was. Maybe he can add some input here.

Regardless of if it is Galley or pilot rejecting there are still issues:
* Just because a namespace defines a ServiceEntry for `google.com` doesn't mean it owns `google.com`. Now if some other namespace wants to call google.com, they need to take a dependency on some other namespaces `google.com` which can be removed at any time
* The "first rule wins" also has the problem that if you update the ServiceEntry then the creationtimestamp will change, so now a difference namespace can steal your `google.com` forever

**Comment by hzxuzhonghu** [comment]
> Just because a namespace defines a ServiceEntry for google.com doesn't mean it owns google.com. Now if some other namespace wants to call google.com, they need to take a dependency on some other namespaces google.com which can be removed at any time


Indeed, we can reject SE creation in galley when there are host conflict in ServiceEntries. But if one is private and one is public exported, then allow it.  Select the private SE first.

> The "first rule wins" also has the problem that if you update the ServiceEntry then the creationtimestamp will change, so now a difference namespace can steal your google.com forever

The creationtimestamp will not change if SE updated.  Also by rejecting conflicting SEs, this will not happen.

**Comment by andraxylia** [comment]
On a related note, shall we remove from the API for virtual_service the possibility to specify short names and infer the namespace? @louiscryan @rshriram . I can envision this causing confusion:

https://github.com/istio/api/blob/master/networking/v1alpha3/virtual_service.proto#L208

**Comment by nrjpoddar** [comment]
@andraxylia there are few other places too where short names are used like DestinationRules and AuthPolicy (AuthPolicy only allows short names). Are we going to remove the short name specification from those resources too?

**Comment by andraxylia** [comment]
@nrjpoddar There is already an inconsistency in DestinationRule, it allows to specify a single host instead of repeated hosts. 

https://github.com/istio/api/blob/master/networking/v1alpha3/destination_rule.proto#L115

That field should match ServiceEntry and VirtualHost.

**Comment by uk000** [comment]
A few notes based on the discussion so far:

- Agree with @howardjohn 's assessment that one service/namespace shouldn't be forced to depend on another namespace just because that other namespace was the first one to grab an FQDN via ServiceEntry.
- Rejecting a ServiceEntry doesn't sound like a reasonable option. As a user of a multi-tenant cluster, I might be completely unaware of what others have in their namespaces, and I'd be confused (and sad) if my changes are rejected by galley or pilot just because some other namespace beat me to it.
- On the aspect of "first rule wins" and "creationtimestamp" , the observed behavior as I noted in my original issue comment was that the 2nd ServiceEntry just keeps waiting for the right time and then hijacks the traffic. Sometimes the second ServiceEntry didn't take effect on the existing client pods, but did take effect on the new client pods. Not sure if it was due to timestamp change or something else.

In general, I'd think that the behavior of a ServiceEntry from one namespace auto-injecting a host mapping in another namespace is counter-intuitive. The argument that "a service should only have one definition in the registry" applies more to kubernetes internal services. A ServiceEntry doesn't truly define a service as much as it's a reference entry (pointer) for a service living elsewhere. From that viewpoint, the mapping, as pointed out by @louiscryan, should be map[hosnameFQDN, namespace-> service] (or even map[namespace, hosnameFQDN-> service]), but just map[hostnameFQDN->service] is too imposing.

A namespace's own ServiceEntries should be the only "addressable" FQDNs introduced into that namespace. I doubt if any user will think like this: "I need to define an FQDN mapping to some destination endpoint, let me check what other namespaces are offering and if some of those can be reused". Instead, one would want to be one's namespace to be completely self-sufficient, and not subject to undefined behavior based on whether some other namespace was added/deleted/modified.

Each namespace should have the right to define its own view of the FQDNs it uses/depends on, including what endpoints it leads to. Even if it means that multiple namespaces might end up defining the same "common" FQDNs again and again, I'd think that to be more intuitive and easy to understand than the cross-namespace dependencies.

**Comment by tiurikov** [comment]
Unfortunately I have no opportunity to contribute, but it would be really nice to have it implemented. 
Is there any plan to fix it in the near future? Any estimations? Thank you!

**Comment by howardjohn** [comment]
@tiurikov this is a top priority for us and I am working on a design doc for this. Will be ready to share in a couple days

**Comment by iandyh** [comment]
@howardjohn Off topic but somehow related. It's also very easy to "hijack" the ingress traffic as well. In the bookinfo example, when other namespace define the same `hosts` in `virtualservice` and `gateway`, by adding other url prefix rules, all the traffic that is supposed to belonged to original namespace will be going the service in other namespaces. 

Hopefully I explained it well, if required, I can open another issue and discuss further.

**Comment by tiurikov** [comment]
I might be wrong, but looks like this issue: https://github.com/istio/istio/issues/9784 has the same root.

**Comment by rshriram** [comment]
btw, another issue here is the multicluster with gateway connectivity thing. In that scenario, we add service entries to the system and use coreDNS to prove DNS resolution for these service entries [based on the address in the service entry object]. If we now allow two service entries in different namespaces, then a dns service like coreDNS will have no way of determining which service entry to use. Not to mention the fact that it's translation tables will be corrupted. 
One might say that in multicluster scenarios, you should not have same service entries in multiple namespaces. But that seems partial. If I come back with a multi-tenant multi-cluster setup, then simply supporting service entries in multiple namespaces will turn off an entire feature.

So tread carefully on whatever decision you choose to make.

btw, @uk000 service entry is not just a reference pointer to some external service. We have a service entry type called INTERNAL which literally refers to internal services, with istio mtls, etc. This is used in multicluster, mesh expansions, and so on.

If we were to scope this solution to only service entries of type external, and fix the GetProxyServiceInstances to return nil if the proxy does not belong to any "internal" service, then the fix that @howardjohn currently has will work, as it handles the outbound cases.

**Comment by tiurikov** [comment]
@howardjohn are there any news regarding a possible solution? What are the chances to get it fixed in 1.1.9 or 1.2?

**Comment by howardjohn** [comment]
@tiurikov please see https://docs.google.com/document/d/15-PU9O22Pb0qTzCfwK2hjwNTcb-CqFuw8JnIgwK3EIM for the design doc. You may need to follow the steps under drive on https://istio.io/about/community/join/ to get access
